matlab_simulation/05-estimation/fastSLAM/func_fastSLAM2.m

```diff
@@ -0,0 +1,80 @@
+%% Fast SLAM2.0 function
+% INPUTS:(in order of call)
+% -------
+% Robot's Current Pose
+% Control Input
+% Motion Noise
+% Measurement Noise
+% Old particle Set
+% Mean of all observed Features
+% covariance of all Observed Features
+% Maximum Range
+% Maximum Field of View
+% time Interval
+% total No of Particles 
+% A vector with flags for observed and new features
+% Set of weights or association index
+% OUTPUTS:(in order of call)
+% --------
+% new Pose of the Robot
+% Measurement Matrix
+% New mean of The particle set
+% New covariance of particles 
+% new ParticleSet
+% A vector containing all the measured indexs
+% Updated vector with flags for newly observed features
+% New Set of Weights or association Index
+% % AUTHOR: BISMAYA SAHOO, EMAIL:bsahoo@uwaterloo.ca
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%NOTE: The difference with FASTSLAM 1.0 is the sample_distribution before
+%the observations and the way the weights are update. 
+
+function[newPose,y,muParticleNew,covParticleNew,newParticleSet,meas_ind,newfeature,w_new]...
+        =fastSLAM2(oldPose,u,MotionNoise,MeasurementNoise,map,particleSet,muFeatOld,covFeatOld,rmax,thmax,timeInterval,totalParticles,newfeature,w)
+    R=MeasurementNoise;Q=MotionNoise;noFeatures=size(map,2);covFeatPred=covFeatOld;muFeatPred=muFeatOld;
+    %Move the robot forward.This Motion has noise Q
+    newPose=motionUpdate_fs(oldPose,u,Q,timeInterval);
+    %Look for features that are in FOV
+    meas_ind=featureSearch_fs(noFeatures,map,newPose,rmax,thmax); 
+    %Take Measurements and accumulate them in a observation matrix
+    y=[];
+    for feature_index = meas_ind       
+        y_new=measurementUpdate_fs(map(:,feature_index),newPose,R);
+        y=[y,y_new];     
+    end
+    %Loop through all particles
+    for particle=1:totalParticles
+       %Move the Particles forward
+       particleSetP(:,particle)=motionUpdate_fs(particleSet(:,particle),u,Q,timeInterval);
+       clear yw hmuw Qw;
+       %Run Individual EKFs for each particles to predict and correct features
+       for j=1:length(meas_ind)
+           i = meas_ind(j);
+           
+%          %Sample Proposal (new in 2.0)
+           [mu_pose,cov_pose,H_pose,H_feat,pred_obs]=sample_proposal(...
```

Comment:
Good work. func_fastSLAM & func_fastSLAM2 only slightly different. on this function call and Qw update. It's better to merge these two function into one and user can select the method by flag. That would be awesome.